friktion/friktion_swap_client/swap.py

```diff
@@ -0,0 +1,521 @@
+        ul_factor = await self._get_token_norm_factor(options_contract.underlying_mint)
+        quote_factor = await self._get_token_norm_factor(options_contract.quote_mint)
+        strike_price = (ul_factor / quote_factor) * options_contract.quote_amount / options_contract.underlying_amount if options_contract.is_call else (quote_factor / ul_factor) * options_contract.underlying_amount / options_contract.quote_amount
+        return {
+            'underlyingAsset': options_contract.underlying_mint,
+            'expiryTimestamp': options_contract.expiry_ts,
+            'isPut': not options_contract.is_call,
+            'strikeAsset': options_contract.quote_mint,
+            'strikePrice': strike_price,
+            'collateralAsset': options_contract.underlying_mint
+        }
+
+    async def _get_token_norm_factor(self, mint: PublicKey) -> int:
+        client = AsyncClient(self.url)
+        await client.is_connected()
+        
+        give_token = AsyncToken(
+            client,
+            mint,
+            TOKEN_PROGRAM_ID,
+            Keypair.generate()
+        )
+
+        return 10 ** (await give_token.get_mint_info()).decimals
+
+    async def get_options_contract_for_key(self, key: PublicKey) -> OptionsContract:
+        client = AsyncClient(self.url)
+        res = await client.is_connected()
+        
+        acc = await OptionsContract.fetch(client, key)
+
+        if acc is None:
+            raise ValueError("No options contract found for key = ", key)
+
+        await client.close()
+        return acc
+
+    async def get_swap_order_for_key(self, key: PublicKey) -> SwapOrder:
+        client = AsyncClient(self.url)
+        res = await client.is_connected()
+        
+        acc = await SwapOrder.fetch(client, key)
+
+        if acc is None:
+            raise ValueError("No swap order found for key = ", key)
+            
+        await client.close()
+        return acc
+
+    async def get_swap_order(self, user: PublicKey, order_id: int) -> SwapOrder:
+        [addr, bump] = find_swap_order_address(user, order_id)
+        
+        acc = await self.get_swap_order_for_key(addr)
+        if acc is None:
+            raise ValueError("No swap found for user = ", user, ', order id = ', order_id)
+
+        return acc
+
+
+    async def get_offer_details(self, user: PublicKey, order_id: int) -> Offer:
+        """
+        Method to get offer details
+        Args:
+            offer_id (int): Offer ID
+        Raises:
+            ValueError: The argument is not a valid offer
+        Returns:
+            details (SwapOrder): Offer details
+        """
+
+        acc = await self.get_swap_order(user, order_id)
+        print('acc = ', acc)
+        return Offer.from_swap_order(acc, find_swap_order_address(user, order_id)[0])
+
+    async def validate_bid(self, wallet: Wallet, bid_details: BidDetails, swap_order: SwapOrder, offer: Offer) -> dict:
```

Comment:
After simplifying `verify_allowance` method (as suggested above), the Wallet instance would not long be necessary here. It could receive only the wallet pub key (or read it from the BidDetails)
On top of that, could we fetch `swap_order` and `offer` within the the method?
So we only parameter needed would the `BidDetails` itself:


```suggestion
    async def validate_bid(self, bid_details: BidDetails) -> dict:
        wallet_address: PublicKey = bid_details.swap_order_owner
        order_id: int = bid_details.order_id
        offer: Offer = await c.get_offer_details(wallet_address, order_id)
        swap_order: SwapOrder = await c.get_swap_order(wallet_address, order_id)
        # ....
```

easing calling the method:
```
wallet_address = PublicKey('7yeeGWxusjqvjYGMSXuyYAdjqr13rJVL9WK2f2u7KbWV')
bid_details =  BidDetails(
    swap_order_owner=wallet_address,
    order_id=swap_order_pre_fill.order_id,
    counterparty_give_pool=creator_give_pool_key,
    counterparty_receive_pool=creator_receive_pool_key,
    bid_size=1,
    bid_price=1,
)
await c.validate_bid(bid_details)
```